### PR TITLE
Initial GMainLoop/GMainContext bindings

### PR DESCRIPTION
**Pull request by sdroege**
https://github.com/gtk-rs/glib/issues/168

### CODE REVIEW

File: src/source.rs [hunk 1]
```diff
@@ -146,3 +195,137 @@ pub fn source_remove(source_id: Id) {
...
+pub const PRIORITY_DEFAULT: Priority = Priority(glib_ffi::G_PRIORITY_DEFAULT);
+pub const PRIORITY_HIGH_IDLE: Priority = Priority(glib_ffi::G_PRIORITY_HIGH_IDLE);
+pub const PRIORITY_DEFAULT_IDLE: Priority = Priority(glib_ffi::G_PRIORITY_DEFAULT_IDLE);
+pub const PRIORITY_LOW: Priority = Priority(glib_ffi::G_PRIORITY_LOW);
+
+/// Adds a closure to be called by the main loop the return `Source` is attached to when it's idle.
+///
+/// `func` will be called repeatedly until it returns `Continue(false)`.
+pub fn idle_source_new<'a, N: Into<Option<&'a str>>, F>(name: N, priority: Priority, func: F) -> Source
+where F: FnMut() -> Continue + Send + 'static {
```
Comment (sdroege) [inline src/source.rs]: The question here is mostly whether we want single functions for creating Sources, or create something with a builder pattern here. Single functions seems more simple to me

**Comment by sdroege** [comment]
For MainContext::invoke() this could in theory be this, but I'm not sure how to handle `FnOnce` inside a `Box<_>`. What am I missing?
This fails with `error[E0161]: cannot move a value of type std::ops::FnOnce() + std::marker::Send: the size of std::ops::FnOnce() + std::marker::Send cannot be statically determined`

```Rust
impl MainContext {
    pub fn invoke<F>(&self, func: F)
    where F: FnOnce() + Send + 'static {
        unsafe {
            glib_ffi::g_main_context_invoke_full(self.to_glib_none().0, glib_ffi::G_PRIORITY_DEFAULT_IDLE, Some(trampoline),
                into_raw(func), Some(destroy_closure))
        }
    }

    pub fn invoke_with_priority<F>(&self, priority: Priority, func: F)
    where F: FnOnce() + Send + 'static {
        unsafe {
            glib_ffi::g_main_context_invoke_full(self.to_glib_none().0, priority.to_glib(), Some(trampoline),
                into_raw(func), Some(destroy_closure))
        }
    }
}

unsafe extern "C" fn trampoline(func: gpointer) -> gboolean {
    let _guard = CallbackGuard::new();
    let func: &RefCell<Option<Box<FnOnce() + Send + 'static>>> = transmute(func);

    if let Some(func) = func.borrow_mut().take() {
        func();
    }

    glib_ffi::G_SOURCE_REMOVE
}

unsafe extern "C" fn destroy_closure(ptr: gpointer) {
    let _guard = CallbackGuard::new();
    Box::<RefCell<Option<Box<FnOnce() + Send + 'static>>>>::from_raw(ptr as *mut _);
}

fn into_raw<F: FnOnce() + Send + 'static>(func: F) -> gpointer {
    let func: Box<RefCell<Option<Box<FnOnce() + Send + 'static>>>> =
        Box::new(RefCell::new(Some(Box::new(func))));
    Box::into_raw(func) as gpointer
}
```

**Comment by sdroege** [comment]
Problem is that we need something like `FnBox` https://doc.rust-lang.org/alloc/boxed/trait.FnBox.html

**Comment by sdroege** [comment]
Good for review then IMHO, @GuillaumeGomez @EPashkin :) No further additions planned to this pull request

File: src/main_context.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    // FIXME: These can actually be FnOnce but require FnBox to
+    // stabilize, or Box<FnOnce()> to be callable otherwise
+    pub fn invoke<F>(&self, func: F)
+    where F: FnMut() + Send + 'static {
+        unsafe {
+            glib_ffi::g_main_context_invoke_full(self.to_glib_none().0, glib_ffi::G_PRIORITY_DEFAULT_IDLE, Some(trampoline),
+                into_raw(func), Some(destroy_closure))
+        }
+    }
```
Comment (EPashkin) [inline src/main_context.rs]: We always pass to FFI boxed closure. And `destroy_closure` working with boxed.

Comment (EPashkin) [inline src/main_context.rs]: Hm. Missed that you added your version `destory_closure`

File: src/lib.rs [hunk 3]
```diff
@@ -183,7 +190,7 @@ mod file_error;
 mod key_file;
 pub mod prelude;
 pub mod signal;
-mod source;
+pub mod source;
```
Comment (EPashkin) [inline src/lib.rs]: Part of source reexported, while part is not.
Not better do `pub use source::*;` ?

**Comment by EPashkin** [comment]
Code looks good to me, but PR missing update Gir submodule and regen to correct version.

File: src/auto/functions.rs [hunk 4]
```diff
@@ -572,10 +569,6 @@ pub fn hostname_to_unicode(hostname: &str) -> Option<String> {
 //    unsafe { TODO: call ffi::g_idle_remove_by_data() }
 //}
 
-//pub fn idle_source_new() -> /*Ignored*/Option<Source> {
```
Comment (GuillaumeGomez) [inline src/auto/functions.rs]: Since we bound `Source`, why is this ignored?

File: src/auto/functions.rs [hunk 5]
```diff
@@ -142,10 +143,6 @@ pub fn check_version(required_major: u32, required_minor: u32, required_micro: u
 //    unsafe { TODO: call ffi::g_child_watch_add_full() }
 //}
 
-//pub fn child_watch_source_new(pid: Pid) -> /*Ignored*/Option<Source> {
```
Comment (GuillaumeGomez) [inline src/auto/functions.rs]: Since we bound `Source`, why is this ignored? (We don't have `Pid`?)

File: src/auto/functions.rs (hunk 4 above)
Comment (GuillaumeGomez) [inline src/auto/functions.rs]: Ah nevermind...

File: src/lib.rs (hunk 3 above)
Comment (sdroege) [inline src/lib.rs]: Your call :)

File: src/source.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/source.rs]: Either way seems fine to me so do whatever seems better to you.

File: src/lib.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/lib.rs]: No preference for me so do whatever you want.

**Comment by GuillaumeGomez** [comment]
Looks good to me as well. Fix @EPashkin's issues and let's merge!

**Comment by sdroege** [comment]
@EPashkin What do you want to have fixed? Should source::* be re-exported at the crate level?

**Comment by EPashkin** [comment]
@sdroege Yes + gir submodule update +  regen

**Comment by sdroege** [comment]
OK, will do in an hour or two

**Comment by sdroege** [comment]
Done :)

**Comment by sdroege** [comment]
Ah I guess that's the reason for not re-exporting source::*. source::Id should be re-exported as SourceId. What do you suggest? Rename Id to SourceId?

**Comment by sdroege** [comment]
Let's try that...

**Comment by EPashkin** [comment]
👍
Thanks @sdroege

**Comment by GuillaumeGomez** [comment]
The "changes PR" is getting really huge. Next release will be certainly a worldwide ovation. People will throw money at us. Gods will sing our names and we'll be taken as examples by future generations. Can't wait to be at the release. :sob:

**Comment by sdroege** [comment]
I'd like to get some more things into the next release though, most importantly removing generic impls for the various conversions traits (and move them into the macros) and implement them for enums/flags/boxed/etc types (which is not possible right now. so we can have those in GValues, for example)

**Comment by GuillaumeGomez** [comment]
I don't intend to make the release before at least another month. I still need to work on gtk-rs/release anyway.

**Comment by sdroege** [comment]
🎉

**Comment by GuillaumeGomez** [comment]
Thanks a lot!

**Pull request merged**